File: backend/model/src/main/java/pl/tiguarces/book/dto/response/BookReactionsResponse.java
package pl.tiguarces.book.dto.response;

import org.springframework.data.domain.Page;
import pl.tiguarces.book.entity.UserBookReaction;

import static pl.tiguarces.AppUtils.formatDate;

public record BookReactionsResponse(Page<Reaction> page, Reaction userReaction, Statistics statistics) {

    public record Statistics(int numberOfOneStars, int numberOfTwoStars,
                             int numberOfThreeStars, int numberOfFourStars,
                             int numberOfFiveStars, int numberOfComments,
                             int numberOfStars) { }

    public record Reaction(long reactionId, int numberOfStars, String comment, String date,
                           String userUsername, String userAvatar) {
        public static Reaction map(final UserBookReaction reaction) {
            var user = reaction.getUser();
            return new Reaction(reaction.getReactionId(), reaction.getNumberOfStars(), reaction.getComment(), formatDate(reaction.getDate()),
                                user.getUsername(), user.getAvatar());
        }
    }
}

File: frontend/src/app/model/BookReactionsResponse.ts
export interface BookReactionsResponse {
  page: Page,
  userReaction: Reaction,
  statistics: Statistics
}

export interface Page {
  content: Reaction[],
  last: boolean,
  totalPages: number,
  size: number,
  first: boolean,
  numberOfElements: number,
  totalElements: number,
  pageable: {
    pageNumber: number
  }
}

export interface Reaction {
  reactionId: number,
  numberOfStars: number,
  comment: string,
  date: string,
  userUsername: string,
  userAvatar: string
}

export interface Statistics {
  numberOfOneStars: number,
  numberOfTwoStars: number,
  numberOfThreeStars: number,
  numberOfFourStars: number,
  numberOfFiveStars: number,
  numberOfComments: number,
  numberOfStars: number
}
